source/game/system/KPadController.cc

```diff
@@ -147,6 +147,59 @@ void RaceInputState::reset() {
     trickRaw = 0;
 }
 
+/// @brief Checks if the input state is valid.
+/// @return If the input state is valid.
+bool RaceInputState::isValid() const {
+    // Check if there are any invalid buttons
+    // We cannot validate from the previous input state because it doesn't exist here
+    if (!!(buttons & ~0xf)) {
+        return false;
+    }
+
+    // Check if the sticks are in the domain
+    // The set of valid stick values is {(n - 7) / 7 | 0 <= n <= 14, in Z}
+    // It's possible for the stick input to be 8/7 with n = 15, but only with ghost controllers
+    if (stick.x > 1.0f || stick.y > 1.0f) {
+        return false;
+    }
+
+    for (size_t i = 0; i <= 14; ++i) {
+        auto xCond = stick.x <=> (static_cast<f32>(i) - 7.0f) / 7.0f;
```

Comment:
I think some parentheses could help. It's difficult to tell if the spaceship or the division comes first (without knowing the order of operations).

Reply:
Suppose we compare integers `x > y` in PowerPC. We generate the `cmpwi rX, rY; ble outer_path` instructions. You can think of the spaceship operator as an isolation of `cmpwi rX, rY`. This allows the compiler to dispatch branch instructions off of one comparison: for example, `cmpwi rX, rY; blt path_less; beq path_equal; bgt path_greater` (if one of these paths were not made implicit).

In this usecase, there are four possible results of float comparison: equivalent, less, greater, or unordered. Unordered means one or both of the operands are NaNs, and are thus incomparable. The compiler will typically generate guards against NaN comparisons automatically, leaving a similar piece of code that doesn't use the spaceship operator to look like:
```x86asm
; assert
UCOMISS XMM1,XMM0
JP assert_path

; comparisons
COMISS XMM1,XMM0
JP continue_path ; this is provably unnecessary, but the compiler forces our hand
JZ break_path
JA continue_path
```
But with the spaceship operator, because we can dispatch all of the branch paths taken, these instructions are much cleaner:
```x86asm
; assert and comparisons are merged
COMISS XMM1,XMM0
JP assert_path
JZ break_path
JBE return_path
; implicit fallthrough into the loop body, which happens to be continue
```